### PR TITLE
CCv0: Merge main into CCv0 branch

### PR DESCRIPTION
**Pull request by Megan-Wright**
Merge remote-tracking branch 'upstream/main' into CCv0
    
Fixes: #6504
Signed-off-by: Megan Wright <megan.wright@ibm.com>

### CODE REVIEW

**Comment by Megan-Wright** [comment]
```
 git diff --stat upstream/main -- . ':!*/vendor/*'
 .github/workflows/cc-payload-after-push-amd64.yaml                            |  159 ++
 .github/workflows/cc-payload-after-push-s390x.yaml                            |  153 ++
 .github/workflows/cc-payload-after-push.yaml                                  |   39 +
 .github/workflows/cc-payload-amd64.yaml                                       |  142 ++
 .github/workflows/cc-payload-s390x.yaml                                       |  134 ++
 .github/workflows/cc-payload.yaml                                             |   39 +
 .github/workflows/commit-message-check.yaml                                   |    4 +-
 .github/workflows/deploy-ccv0-demo.yaml                                       |  124 ++
 .github/workflows/kata-deploy-push.yaml                                       |    4 +
 .github/workflows/static-checks.yaml                                          |    4 +-
 ci/install_libseccomp.sh                                                      |    3 +-
 ci/lib.sh                                                                     |   83 +
 docs/how-to/README.md                                                         |    3 +
 docs/how-to/ccv0.sh                                                           |  640 ++++++
 docs/how-to/data/confidential-agent-config.toml.in                            |   45 +
 docs/how-to/how-to-build-and-test-ccv0.md                                     |  475 ++++
 ...rate-a-kata-containers-payload-for-the-confidential-containers-operator.md |   44 +
 docs/how-to/how-to-set-sandbox-config-kata.md                                 |   10 +
 docs/how-to/how-to-setup-swap-devices-in-guest-kernel.md                      |    2 +
 docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md                           |    1 +
 docs/how-to/how-to-use-virtio-mem-with-kata.md                                |    2 +
 src/agent/Cargo.lock                                                          | 4077 ++++++++++++++++++++++++++++++-----
 src/agent/Cargo.toml                                                          |   14 +-
 src/agent/rustjail/src/cgroups/fs/mod.rs                                      |   19 +-
 src/agent/rustjail/src/lib.rs                                                 |    8 +-
 src/agent/src/config.rs                                                       |  220 +-
 src/agent/src/device.rs                                                       |   18 +-
 src/agent/src/image_rpc.rs                                                    |  337 +++
:...skipping...
 .github/workflows/cc-payload-after-push-amd64.yaml                            |  159 ++
 .github/workflows/cc-payload-after-push-s390x.yaml                            |  153 ++
 .github/workflows/cc-payload-after-push.yaml                                  |   39 +
 .github/workflows/cc-payload-amd64.yaml                                       |  142 ++
 .github/workflows/cc-payload-s390x.yaml                                       |  134 ++
 .github/workflows/cc-payload.yaml                                             |   39 +
 .github/workflows/commit-message-check.yaml                                   |    4 +-
 .github/workflows/deploy-ccv0-demo.yaml                                       |  124 ++
 .github/workflows/kata-deploy-push.yaml                                       |    4 +
 .github/workflows/static-checks.yaml                                          |    4 +-
 ci/install_libseccomp.sh                                                      |    3 +-
 ci/lib.sh                                                                     |   83 +
 docs/how-to/README.md                                                         |    3 +
 docs/how-to/ccv0.sh                                                           |  640 ++++++
 docs/how-to/data/confidential-agent-config.toml.in                            |   45 +
 docs/how-to/how-to-build-and-test-ccv0.md                                     |  475 ++++
 ...rate-a-kata-containers-payload-for-the-confidential-containers-operator.md |   44 +
 docs/how-to/how-to-set-sandbox-config-kata.md                                 |   10 +
 docs/how-to/how-to-setup-swap-devices-in-guest-kernel.md                      |    2 +
 docs/how-to/how-to-use-virtio-fs-nydus-with-kata.md                           |    1 +
 docs/how-to/how-to-use-virtio-mem-with-kata.md                                |    2 +
 src/agent/Cargo.lock                                                          | 4077 ++++++++++++++++++++++++++++++-----
 src/agent/Cargo.toml                                                          |   14 +-
 src/agent/rustjail/src/cgroups/fs/mod.rs                                      |   19 +-
 src/agent/rustjail/src/lib.rs                                                 |    8 +-
 src/agent/src/config.rs                                                       |  220 +-
 src/agent/src/device.rs                                                       |   18 +-
 src/agent/src/image_rpc.rs                                                    |  337 +++
 src/agent/src/main.rs                                                         |    1 +
 src/agent/src/rpc.rs                                                          |  323 ++-
 src/agent/src/sandbox.rs                                                      |    2 +
 src/agent/src/signal.rs                                                       |    2 +-
 src/dragonball/src/api/v1/instance_info.rs                                    |   20 +-
 src/dragonball/src/api/v1/mod.rs                                              |    2 +-
 src/dragonball/src/error.rs                                                   |    4 +
 src/dragonball/src/kvm_context.rs                                             |   47 +-
 src/dragonball/src/vm/mod.rs                                                  |   30 +-
 src/libs/oci/src/lib.rs                                                       |   10 +-
 src/libs/protocols/.gitignore                                                 |    3 +
 src/libs/protocols/build.rs                                                   |   21 +-
 src/libs/protocols/protos/image.proto                                         |   31 +
 src/libs/protocols/src/lib.rs                                                 |    4 +
 src/libs/protocols/src/trans.rs                                               |    4 +-
 src/runtime/Makefile                                                          |  137 ++
 src/runtime/arch/amd64-options.mk                                             |    2 +
 src/runtime/cmd/kata-runtime/kata-check_amd64.go                              |    2 +
 src/runtime/cmd/kata-runtime/kata-check_arm64.go                              |    3 +
 src/runtime/cmd/kata-runtime/kata-check_ppc64le.go                            |    3 +
 src/runtime/cmd/kata-runtime/kata-check_s390x.go                              |    3 +
 src/runtime/config/configuration-clh-tdx.toml.in                              |  443 ++++
 src/runtime/config/configuration-clh.toml.in                                  |    2 +-
 src/runtime/config/configuration-qemu-se.toml.in                              |  673 ++++++
 src/runtime/config/configuration-qemu-sev.toml.in                             |  672 ++++++
 src/runtime/config/configuration-qemu-tdx.toml.in                             |  651 ++++++
 src/runtime/config/configuration-qemu.toml.in                                 |   30 +-
 src/runtime/go.mod                                                            |    3 +-
 src/runtime/go.sum                                                            |  282 +--
 src/runtime/hack/update-generated-hypervisor-proto.sh                         |   13 +
 src/runtime/pkg/README.md                                                     |    1 +
 src/runtime/pkg/containerd-shim-v2/create.go                                  |    5 +
 src/runtime/pkg/containerd-shim-v2/image.go                                   |   84 +
 src/runtime/pkg/govmm/qemu/qemu.go                                            |   45 +-
 src/runtime/pkg/govmm/qemu/qmp.go                                             |   66 +
 src/runtime/pkg/katautils/config-settings.go.in                               |    7 +
 src/runtime/pkg/katautils/config.go                                           |  285 ++-
 src/runtime/pkg/katautils/config_test.go                                      |    2 +
 src/runtime/pkg/oci/utils.go                                                  |   32 +
 src/runtime/pkg/oci/utils_test.go                                             |    2 +-
 src/runtime/pkg/sev/README.md                                                 |   14 +
 src/runtime/pkg/sev/kbs/kbs.go                                                |   33 +
 src/runtime/pkg/sev/ovmf.go                                                   |  101 +
 src/runtime/pkg/sev/sev.go                                                    |  203 ++
 src/runtime/pkg/sev/sev_test.go                                               |   54 +
 src/runtime/pkg/sev/testdata/README.md                                        |    9 +
 src/runtime/pkg/sev/testdata/ovmf_suffix.bin                                  |  Bin 0 -> 4096 bytes
 src/runtime/pkg/sev/vcpu_sigs.go                                              |   76 +
 src/runtime/pkg/sev/vcpu_sigs_test.go                                         |   21 +
 src/runtime/pkg/sev/vmsa.go                                                   |  172 ++
 src/runtime/protocols/hypervisor/hypervisor.pb.go                             | 1848 ++++++++++++++++
 src/runtime/protocols/hypervisor/hypervisor.proto                             |   46 +
 src/runtime/protocols/simple-kbs/keybroker.pb.go                              |  573 +++++
 src/runtime/protocols/simple-kbs/keybroker.proto                              |   68 +
 src/runtime/protocols/simple-kbs/keybroker_grpc.pb.go                         |  141 ++
 src/runtime/virtcontainers/acrn.go                                            |    9 +
 src/runtime/virtcontainers/agent.go                                           |    3 +
 src/runtime/virtcontainers/clh.go                                             |   15 +-
 src/runtime/virtcontainers/container.go                                       |    5 +
 src/runtime/virtcontainers/fc.go                                              |    4 +
 src/runtime/virtcontainers/fs_share_linux.go                                  |   57 +-
 src/runtime/virtcontainers/hypervisor.go                                      |  443 +---
 src/runtime/virtcontainers/hypervisor_config_darwin.go                        |    4 +
 src/runtime/virtcontainers/hypervisor_config_linux.go                         |    4 +
 src/runtime/virtcontainers/hypervisor_linux.go                                |    2 +
 src/runtime/virtcontainers/image/image.go                                     |   24 +
 src/runtime/virtcontainers/interfaces.go                                      |    3 +
 src/runtime/virtcontainers/kata_agent.go                                      |  222 +-
 src/runtime/virtcontainers/mock_agent.go                                      |    5 +
 src/runtime/virtcontainers/mock_hypervisor.go                                 |    4 +
 src/runtime/virtcontainers/monitor.go                                         |   11 +-
 src/runtime/virtcontainers/mount.go                                           |   24 +-
 src/runtime/virtcontainers/pkg/agent/protocols/client/client.go               |   35 +
 src/runtime/virtcontainers/pkg/agent/protocols/grpc/image.pb.go               |  613 ++++++
 src/runtime/virtcontainers/pkg/annotations/annotations.go                     |   25 +-
 src/runtime/virtcontainers/pkg/cloud-hypervisor/client/api/openapi.yaml       |    3 -
 src/runtime/virtcontainers/pkg/cloud-hypervisor/client/docs/CpuAffinity.md    |   16 +-
 src/runtime/virtcontainers/pkg/cloud-hypervisor/client/model_cpu_affinity.go  |   64 +-
 src/runtime/virtcontainers/pkg/cloud-hypervisor/cloud-hypervisor.yaml         |    3 -
 src/runtime/virtcontainers/pkg/mock/mock.go                                   |    4 +
 src/runtime/virtcontainers/pkg/vcmock/sandbox.go                              |    5 +
 src/runtime/virtcontainers/qemu.go                                            |  136 +-
 src/runtime/virtcontainers/qemu_amd64.go                                      |  243 ++-
 src/runtime/virtcontainers/qemu_amd64_test.go                                 |   24 +-
 src/runtime/virtcontainers/qemu_arch_base.go                                  |   56 +-
 src/runtime/virtcontainers/remote.go                                          |  298 +++
 src/runtime/virtcontainers/sandbox.go                                         |   76 +-
 src/runtime/virtcontainers/types/sandbox.go                                   |   12 +
 src/tools/agent-ctl/src/client.rs                                             |  117 +-
 src/tools/agent-ctl/src/utils.rs                                              |    2 +-
 tools/osbuilder/image-builder/Dockerfile                                      |    1 +
 tools/osbuilder/image-builder/image_builder.sh                                |   22 +-
 tools/osbuilder/rootfs-builder/README.md                                      |    2 +-
 tools/osbuilder/rootfs-builder/agent-config.toml.in                           |   44 +
 tools/osbuilder/rootfs-builder/alpine/Dockerfile.in                           |    4 +
 tools/osbuilder/rootfs-builder/clearlinux/Dockerfile.in                       |    5 +
 tools/osbuilder/rootfs-builder/init_trusted_storage.sh                        |  130 ++
 tools/osbuilder/rootfs-builder/rootfs.sh                                      |   63 +
 tools/osbuilder/rootfs-builder/ubuntu/Dockerfile.in                           |   26 +-
 tools/osbuilder/rootfs-builder/ubuntu/config.sh                               |   21 +-
 tools/osbuilder/rootfs-builder/ubuntu/rootfs_lib.sh                           |   28 +-
 tools/osbuilder/scripts/lib.sh                                                |    9 +
 tools/packaging/guest-image/build_image.sh                                    |   60 +-
 tools/packaging/guest-image/build_se_image.sh                                 |   96 +
 tools/packaging/kata-deploy-cc/Dockerfile                                     |   30 +
 tools/packaging/kata-deploy-cc/scripts/kata-deploy.sh                         |  288 +++
 tools/packaging/kata-deploy/kata-cleanup/base/kata-cleanup.yaml               |    2 +-
 tools/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml                 |    4 +-
 tools/packaging/kata-deploy/local-build/Makefile                              |   75 +
 tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh     |    8 +-
 tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh               |  476 +++-
 .../packaging/kata-deploy/local-build/kata-deploy-build-and-upload-payload.sh |    5 +-
 tools/packaging/kata-deploy/local-build/kata-deploy-merge-builds.sh           |    1 +
 tools/packaging/kata-deploy/scripts/kata-deploy.sh                            |    0
 tools/packaging/kernel/build-kernel.sh                                        |   37 +-
 .../kernel/configs/fragments/common/confidential_containers/cryptsetup.conf   |   21 +
 .../kernel/configs/fragments/common/confidential_containers/initramfs.conf    |    1 +
 tools/packaging/kernel/configs/fragments/x86_64/crypto.conf                   |    3 +
 tools/packaging/kernel/kata_config_version                                    |    2 +-
 tools/packaging/qemu/patches/7.1.x/no_patches.txt                             |    0
 tools/packaging/qemu/patches/tag_patches/v7.1.0/no_patches.txt                |    0
 tools/packaging/scripts/configure-hypervisor.sh                               |    4 +-
 tools/packaging/scripts/lib.sh                                                |   59 +-
 tools/packaging/static-build/cache_components.sh                              |  274 +++
 tools/packaging/static-build/initramfs/Dockerfile                             |   68 +
 tools/packaging/static-build/initramfs/build-initramfs.sh                     |   15 +
 tools/packaging/static-build/initramfs/build.sh                               |   47 +
 tools/packaging/static-build/initramfs/init.sh                                |   44 +
 tools/packaging/static-build/initramfs/initramfs.list                         |   21 +
 tools/packaging/static-build/kernel/Dockerfile                                |    1 +
 tools/packaging/static-build/kernel/build.sh                                  |    1 +
 tools/packaging/static-build/ovmf/Dockerfile                                  |   10 +-
 tools/packaging/static-build/ovmf/build-ovmf.sh                               |    7 +-
 tools/packaging/static-build/ovmf/build.sh                                    |   18 +-
 tools/packaging/static-build/qemu/build-static-qemu-cc.sh                     |   34 +
 tools/packaging/static-build/shim-v2/build.sh                                 |   16 +-
 tools/packaging/static-build/td-shim/Dockerfile                               |    4 +-
 tools/packaging/static-build/td-shim/build-td-shim.sh                         |    4 +-
 tools/packaging/static-build/td-shim/build.sh                                 |    2 +-
 utils.mk                                                                      |    4 +-
 versions.yaml                                                                 |   59 +-
 169 files changed, 17209 insertions(+), 1632 deletions(-)
```

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 1]
```diff
@@ -207,7 +210,7 @@ install_cached_cc_shim_v2() {
 
 # Install static CC cloud-hypervisor asset
 install_cc_clh() {
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:213]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 2]
```diff
@@ -266,7 +269,7 @@ install_cc_image() {
 	local pause_version="$(get_from_kata_deps "externals.pause.version")"
 	local rust_version="$(get_from_kata_deps "languages.rust.meta.newest-version")"
 
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:272]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 3]
```diff
@@ -312,7 +315,7 @@ install_cc_kernel() {
 
 	local kernel_kata_config_version="$(cat ${repo_root_dir}/tools/packaging/kernel/kata_config_version)"
 
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:318]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 4]
```diff
@@ -332,7 +335,7 @@ install_cc_qemu() {
 	export qemu_repo="$(yq r $versions_yaml assets.hypervisor.qemu.url)"
 	export qemu_version="$(yq r $versions_yaml assets.hypervisor.qemu.version)"
 
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:338]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 5]
```diff
@@ -389,7 +392,7 @@ install_cc_shimv2() {
 # Install static CC virtiofsd asset
 install_cc_virtiofsd() {
 	local virtiofsd_version="$(get_from_kata_deps "externals.virtiofsd.version")-$(get_from_kata_deps "externals.virtiofsd.toolchain")"
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:395]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 6]
```diff
@@ -413,7 +416,7 @@ install_cached_kernel_component() {
 
 	local kernel_kata_config_version="$(cat ${repo_root_dir}/tools/packaging/kernel/kata_config_version)"
 
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:419]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 7]
```diff
@@ -481,7 +484,7 @@ install_cc_tee_qemu() {
 	export qemu_version="$(yq r $versions_yaml assets.hypervisor.qemu.${tee}.tag)"
 	export tee="${tee}"
 
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:487]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 8]
```diff
@@ -425,7 +428,7 @@ install_cached_kernel_component() {
 	[ "${tee}" == "tdx" ] && return 0
 
 	# SEV specific code path
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:431]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 9]
```diff
@@ -499,7 +502,7 @@ install_cc_tdx_qemu() {
 }
 
 install_cc_tdx_td_shim() {
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:505]: ... this one ...

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 10]
```diff
@@ -519,7 +522,7 @@ install_cc_tee_ovmf() {
 	local component_name="ovmf"
 	local component_version="$(get_from_kata_deps "externals.ovmf.${tee}.version")"
 	[ "${tee}" == "tdx" ] && component_name="tdvf"
-	install_cached_component \
+	install_cached_tarball_component \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh:525]: ... this one ...

File: tools/packaging/scripts/lib.sh [hunk 11]
```diff
@@ -208,13 +208,13 @@ get_ovmf_image_name() {
 }
 
 get_qemu_image_name() {
-	qemu_script_dir="${this_script_dir}/../static-build/qemu"
-	echo "${CC_BUILDER_REGISTRY}:qemu-$(get_last_modification ${qemu_script_dir})-$(uname -m)"
+	qemu_script_dir="${repo_root_dir}/tools/packaging/static-build/qemu"
+	echo "${BUILDER_REGISTRY}:qemu-$(get_last_modification ${qemu_script_dir})-$(uname -m)"
```
Comment (fidencio) [inline tools/packaging/scripts/lib.sh:212]: We should keep `CC_BUILDER_REGSTRY` for the `CCv0` branch ...

File: tools/packaging/scripts/lib.sh [hunk 12]
```diff
@@ -208,13 +208,13 @@ get_ovmf_image_name() {
...
 }
 
 get_shim_v2_image_name() {
-	shim_v2_script_dir="${this_script_dir}/../static-build/shim-v2"
-	echo "${CC_BUILDER_REGISTRY}:shim-v2-go-$(get_from_kata_deps "languages.golang.meta.newest-version")-rust-$(get_from_kata_deps "languages.rust.meta.newest-version")-$(get_last_modification ${shim_v2_script_dir})-$(uname -m)"
+	shim_v2_script_dir="${repo_root_dir}/tools/packaging/static-build/shim-v2"
+	echo "${BUILDER_REGISTRY}:shim-v2-go-$(get_from_kata_deps "languages.golang.meta.newest-version")-rust-$(get_from_kata_deps "languages.rust.meta.newest-version")-$(get_last_modification ${shim_v2_script_dir})-$(uname -m)"
```
Comment (fidencio) [inline tools/packaging/scripts/lib.sh:217]: ... the same here ...

**Review by fidencio — CHANGES_REQUESTED** [review]
I've left a few comments about the bits that I added to the `main` branch and that shouldn't be considered as part of the `CCv0` one.

**Comment by Megan-Wright** [comment]
@fidencio thanks for spotting those - updated now!

**Comment by fidencio** [comment]
We're hitting a bunch of:
```
11:29:36 warning: spurious network error (2 tries remaining): [92] Stream error in the HTTP/2 framing layer (HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1))
11:29:38 warning: spurious network error (2 tries remaining): [18] Transferred a partial file (transfer closed with 1845588 bytes remaining to read)
11:29:46 warning: spurious network error (2 tries remaining): [18] Transferred a partial file (transfer closed with 667659 bytes remaining to read)
11:29:53 warning: spurious network error (2 tries remaining): [92] Stream error in the HTTP/2 framing layer (HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1))
11:30:09 warning: spurious network error (1 tries remaining): [28] Timeout was reached (download of `libz-sys v1.1.8` failed to transfer more than 10 bytes in 30s)
11:32:43 warning: spurious network error (1 tries remaining): [28] Timeout was reached (failed to download any data for `prost-build v0.10.4` within 30s)
```

I'd like to ignore those and move forward with this PR.

**Pull request merged**